node/node.go

```diff
@@ -452,26 +503,121 @@ func (node *Node) StartSyncingServer() {
 			blockHash := block.Hash()
+			if bytes.Compare(blockHash[:], startHeaderHash) == 0 {
+				break
+			}
 			response.Payload = append(response.Payload, blockHash[:])
 		}
-	} else {
+	case downloader_pb.DownloaderRequest_BLOCK:
 		for _, bytes := range request.Hashes {
 			var hash common.Hash
 			hash.SetBytes(bytes)
 			block := node.blockchain.GetBlockByHash(hash)
 			encodedBlock, err := rlp.EncodeToBytes(block)
-			if err != nil {
+			if err == nil {
 				response.Payload = append(response.Payload, encodedBlock)
 			}
 		}
+	case downloader_pb.DownloaderRequest_NEWBLOCK:
+		if node.State != NodeNotInSync {
+			node.log.Debug("[sync] new block received, but state is", "state", node.State.String())
+			response.Type = downloader_pb.DownloaderResponse_INSYNC
+			return response, nil
+		}
+		var blockObj types.Block
+		err := rlp.DecodeBytes(request.BlockHash, &blockObj)
+		if err != nil {
+			node.log.Warn("[sync] unable to decode received new block")
+			return response, err
+		}
+		node.stateSync.AddNewBlock(request.PeerHash, &blockObj)
+
+	case downloader_pb.DownloaderRequest_REGISTER:
+		peerID := binary.BigEndian.Uint32(request.PeerHash)
+		if _, ok := node.peerRegistrationRecord[peerID]; ok {
+			response.Type = downloader_pb.DownloaderResponse_FAIL
+			return response, nil
+		} else if len(node.peerRegistrationRecord) >= maxBroadcastNodes {
+			response.Type = downloader_pb.DownloaderResponse_FAIL
+			return response, nil
+		} else {
+			peer, ok := node.Consensus.GetPeerFromID(peerID)
+			if !ok {
+				node.log.Warn("[sync] unable to get peer from peerID", "peerID", peerID)
+			}
+			client := downloader.ClientSetup(peer.IP, GetSyncingPort(peer.Port))
+			if client == nil {
+				node.log.Warn("[sync] unable to setup client")
+				return response, nil
+			}
+			node.log.Debug("[sync] client setup correctly", "client", client)
+			config := &syncConfig{timestamp: time.Now().UnixNano(), client: client}
+			node.stateMutex.Lock()
+			node.peerRegistrationRecord[peerID] = config
+			node.stateMutex.Unlock()
+			node.log.Debug("[sync] register peerID success", "peerID", peerID)
+			response.Type = downloader_pb.DownloaderResponse_SUCCESS
+		}
+	case downloader_pb.DownloaderRequest_REGISTERTIMEOUT:
+		if node.State == NodeNotInSync {
+			count := node.stateSync.RegisterNodeInfo()
+			node.log.Debug("[sync] extra node registered", "number", count)
+		}
 	}
 	return response, nil
 }
 
+// SendNewBlockToUnsync send latest verified block to unsync, registered nodes
+func (node *Node) SendNewBlockToUnsync() {
+	for {
+		block := <-node.Consensus.VerifiedNewBlock
+		blockHash, err := rlp.EncodeToBytes(block)
+		if err != nil {
+			node.log.Warn("[sync] unable to encode block to hashes")
+			continue
+		}
+
+		// really need to have a unique id independent of ip/port
+		selfPeerID := utils.GetUniqueIDFromIPPort(node.SelfPeer.IP, node.SelfPeer.Port)
+		peerHash := make([]byte, 4)
```

Comment:
if you need to serialize the peerid, please do it when you are ready to send to network.

Reply:
put it into pushnewblock